### PR TITLE
Spec for GatherElements-6

### PR DESCRIPTION
**Pull request by pavel-esir**
Need to add new operation GatherElements-6 to support Machine Learning models exported to ONNX. GatherElements is a special case of GatherND and can be expressed through it but performance is very poor. For the sake of performance boost this new operation was introduced.

**No new MO transformation will be added for this operation**

tickets: #41387, #42627

### CODE REVIEW

File: docs/ops/movement/GatherElements_6.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        [4, 8, 3],
+      ],
+  ]
+```
+**Attributes**:
+* *axis* 
+  * **Description**: Which axis to gather on. Negative value means counting dimensions from the back. 
+  * **Range of values**: `[-r, r-1]` where `r = rank(data)`.
+  * **Type**: int
+  * **Default value**: 0
```
Comment (andrei-kochin) [inline docs/ops/movement/GatherElements_6.md:83]: - Default value
Required: yes

Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md]: Done

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: docs/ops/movement/GatherElements_6.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  ]
+  axis = 0
+  output = [
+      [
+        [4, 2, 6],
+        [4, 8, 3],
+      ],
+  ]
+```
+**Attributes**:
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md:78]: Done.

File: docs/ops/movement/GatherElements_6.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+**Category**: Data movement operations
+
+**Short description**: *GatherElements* takes elements from input `data` tensor at positions specified in `indices`.
+
+**Detailed description** *GatherElements* takes elements from `data` tensor at positions specified in `indices`. `data` 
+and `indices` tensors have the same rank `r >= 1`. Optional attribute `axis` (by default `axis` is 0) determines 
+along which axis elements with indices specified in `indices` are taken. `indices` tensor has the same shape as `data` 
+except for `axis` dimension. Output consists of values (gathered from the `data`) for each element in `indices` 
+and has the same shape as `indices`. Operation can be expressed through GatherND-5 from opset but in this case execution
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md:13]: Removed.

File: docs/ops/movement/GatherElements_6.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
+## GatherElements <a name="GatherElements"></a> {#openvino_docs_ops_movement_GatherElements_6}
+
+**Versioned name**: *GatherElements-6*
+
+**Category**: Data movement operations
+
+**Short description**: *GatherElements* takes elements from input `data` tensor at positions specified in `indices`.
+
+**Detailed description** *GatherElements* takes elements from `data` tensor at positions specified in `indices`. `data` 
+and `indices` tensors have the same rank `r >= 1`. Optional attribute `axis` (by default `axis` is 0) determines
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md]: removed that phrase

**Review by lazarevevgeny — APPROVED** [review]

File: docs/ops/movement/GatherElements_6.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
+## GatherElements <a name="GatherElements"></a> {#openvino_docs_ops_movement_GatherElements_6}
+
+**Versioned name**: *GatherElements-6*
+
+**Category**: Data movement operations
+
+**Short description**: *GatherElements* takes elements from input `data` tensor at positions specified in `indices`.
```
Comment (lazarevevgeny) [inline docs/ops/movement/GatherElements_6.md:7]: Please, add "the" before all tensor names like: `data`, `indices` etc.

Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md]: Done.

File: docs/ops/movement/GatherElements_6.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        [4, 2, 6],
+        [4, 8, 3],
+      ],
+  ]
+```
+
+**Attributes**:
+* *axis* 
+  * **Description**: Which axis to gather on. Negative value means counting dimensions from the back. 
+  * **Range of values**: `[-r, r-1]` where `r = rank(data)`.
```
Comment (andrejsokolov) [inline docs/ops/movement/GatherElements_6.md:81]: Are we going to support negative values of axis?

Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md:81]: We must be able to read IRs and ONNX models with negative axis, this is why this is in the spec. But we will convert all negative axis to positive ones on the side on ngraph.

Comment (andrejsokolov) [reply docs/ops/movement/GatherElements_6.md:81]: Ok. Thank you for the clarification!

File: docs/ops/movement/GatherElements_6.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      
+* *T*: any supported type.
+
+* *T_IND*: `int32` or `int64`.
+
+**Example**
+
+```xml
+<... type="GatherElements" ...>
+    <data axis=1 />
```
Comment (lazarevevgeny) [inline docs/ops/movement/GatherElements_6.md:108]: axis="1"

File: docs/ops/movement/GatherElements_6.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    <input>
+        <port id="0">
+            <dim>2</dim>
+            <dim>2</dim>
+        </port>
+        <port id="1">
+            <dim>2</dim>
+            <dim>3</dim>
```
Comment (lazarevevgeny) [inline docs/ops/movement/GatherElements_6.md:116]: Please, use different values for the input shape like "3" and "5". And it is better to provide an example with 3D input tensor

File: docs/ops/movement/GatherElements_6.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+* *axis* 
+  * **Description**: Which axis to gather on. Negative value means counting dimensions from the back. 
+  * **Range of values**: `[-r, r-1]` where `r = rank(data)`.
+  * **Type**: int
+  * **Required**: *yes*
+
+
+**Inputs**:
+
+* **1**:  Tensor of type *T*. This is a tensor of a `rank >= 1`. Required.
```
Comment (lazarevevgeny) [inline docs/ops/movement/GatherElements_6.md:88]: \*\*Required\*\*
The same for line 90.

File: docs/ops/movement/GatherElements_6.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+* **2**:  Tensor of type *T_IND* with the same rank as the input. All index values are expected to be within
+ bounds `[0, s-1]`, where `s` is size along `axis` dimension of the `data` tensor. It is an error if any of the index 
+ values are out of bounds. Required.
```
Comment (lazarevevgeny) [inline docs/ops/movement/GatherElements_6.md:92]: `It is an error if any of the index values are out of bounds.`

@dmitry-gorokhov , @vladimir-paramuzov , @taka-no-me may be it is better to write that the behavior is undefined in this case? I believe that the plugin implementations will not have an "if" statement to check that the value is in the boundaries, so the data will be read from undetermined location.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: docs/ops/movement/GatherElements_6.md (hunk 8 above)
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md:115]: Added nontrivial 3D shapes.

File: docs/ops/movement/GatherElements_6.md (hunk 9 above)
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md]: Done.

File: docs/ops/movement/GatherElements_6.md (hunk 10 above)
Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md]: It analogy with GatherND operation. Removed information about error and leaved only info about bounds.

**Review by andrejsokolov — APPROVED** [review]

File: docs/ops/movement/GatherElements_6.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      [0, 1],
+      [0, 0],
+  ]
+  axis = 0
+  output = [
+      [
+        [1, 4],
+        [1, 2],
+      ],
+  ]
```
Comment (nshchego) [inline docs/ops/movement/GatherElements_6.md:37]: Looks like output has shape [1, 2, 2] here, but it should be [2, 2].
The same comments for other examples.

Comment (pavel-esir) [reply docs/ops/movement/GatherElements_6.md:35]: Thanks for noting that. Fixed.

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**